File: flutter_readium/android/src/main/kotlin/dk/nota/flutterreadium/navigators/PageBreakSkippingContentIteratorFactory.kt
package dk.nota.flutterreadium.navigators

import org.readium.r2.shared.ExperimentalReadiumApi
import org.readium.r2.shared.publication.LocalizedString
import org.readium.r2.shared.publication.Locator
import org.readium.r2.shared.publication.Manifest
import org.readium.r2.shared.publication.PublicationServicesHolder
import org.readium.r2.shared.publication.html.cssSelector
import org.readium.r2.shared.publication.services.content.Content
import org.readium.r2.shared.publication.services.content.iterators.HtmlResourceContentIterator
import org.readium.r2.shared.publication.services.content.iterators.ResourceContentIteratorFactory
import org.readium.r2.shared.util.mediatype.MediaType
import org.readium.r2.shared.util.resource.Resource

/**
 * Wraps [HtmlResourceContentIterator.Factory] and handles EPUB page-break elements for TTS.
 *
 * Page-break elements are identified by matching the [Content.TextElement] CSS selector against
 * the fragment IDs declared in the publication's pageList nav. Block-level page-break elements
 * (the typical DAISY/Nordic EPUB3 form) carry the element's own ID as their CSS selector, so
 * the match is exact. Inline spans inside a paragraph are not filtered (their text is merged
 * into the parent block's element and not individually addressable at this level).
 *
 * When [skipPageBreaks] is `true` (default), page-break elements are removed from the iterator.
 * When `false`, they are kept but the label text from the original element is rewritten to a
 * localized string (e.g. "Page 42" or "side 42"), derived from the publication's declared
 * language. [skipPageBreaks] can be changed at any time; the change takes effect on the next
 * [Content.Iterator.hasNext] / [Content.Iterator.hasPrevious] call.
 */
@OptIn(ExperimentalReadiumApi::class)
class PageBreakSkippingContentIteratorFactory(
    var skipPageBreaks: Boolean = true,
    private val delegate: ResourceContentIteratorFactory = HtmlResourceContentIterator.Factory(),
) : ResourceContentIteratorFactory {
    override suspend fun create(
        manifest: Manifest,
        servicesHolder: PublicationServicesHolder,
        readingOrderIndex: Int,
        resource: Resource,
        mediaType: MediaType,
        locator: Locator,
    ): Content.Iterator? {
        val inner =
            delegate.create(manifest, servicesHolder, readingOrderIndex, resource, mediaType, locator)
                ?: return null

        val pageBreakIds = manifest.pageBreakIds()
        if (pageBreakIds.isEmpty()) return inner
        return PageBreakHandlingIterator(inner, pageBreakIds, ::skipPageBreaks, manifest.pageLabel())
    }

    private fun Manifest.pageBreakIds(): Set<String> =
        subcollections["pageList"]
            ?.firstOrNull()
            ?.links
            ?.mapNotNull { link ->
                link.href
                    .toString()
                    .substringAfter("#", "")
                    .takeIf { it.isNotBlank() }
            }?.toSet()
            ?: emptySet()

    private fun Manifest.pageLabel(): ((String) -> String)? {
        val format =
            PAGE_LABEL_FORMATS.getOrFallback(metadata.languages.firstOrNull()) ?: return null
        return { label -> format.string.replace("{label}", label) }
    }

    companion object {
        private val PAGE_LABEL_FORMATS =
            LocalizedString.fromStrings(
                mapOf(
                    "en" to "Page {label}",
                    "da" to "side {label}",
                    "sv" to "sida {label}",
                    "no" to "side {label}",
                    "is" to "{label}. síða",
                ),
            )
    }
}

@OptIn(ExperimentalReadiumApi::class)
private class PageBreakHandlingIterator(
    private val delegate: Content.Iterator,
    private val pageBreakIds: Set<String>,
    private val shouldPageElementsSkip: () -> Boolean,
    private val pageLabel: ((String) -> String)?,
) : Content.Iterator {
    private data class Pending(
        val element: Content.Element,
        val forward: Boolean,
    )

    private var pending: Pending? = null

    override suspend fun hasNext(): Boolean {
        pending?.let { if (it.forward) return true }
        pending = null
        while (delegate.hasNext()) {
            val element = delegate.next()
            if (!shouldPageElementsSkip() || !element.isPageBreak()) {
                pending = Pending(element, forward = true)
                return true
            }
        }
        return false
    }

    override fun next(): Content.Element =
        pending
            ?.takeIf { it.forward }
            ?.element
            ?.also { pending = null }
            ?.transform()
            ?: throw IllegalStateException(
                "Called next() without a successful call to hasNext() first",
            )

    override suspend fun hasPrevious(): Boolean {
        pending?.let { if (!it.forward) return true }
        pending = null
        while (delegate.hasPrevious()) {
            val element = delegate.previous()
            if (!shouldPageElementsSkip() || !element.isPageBreak()) {
                pending = Pending(element, forward = false)
                return true
            }
        }
        return false
    }

    override fun previous(): Content.Element =
        pending
            ?.takeIf { !it.forward }
            ?.element
            ?.also { pending = null }
            ?.transform()
            ?: throw IllegalStateException(
                "Called previous() without a successful call to hasPrevious() first",
            )

    private fun Content.Element.transform(): Content.Element {
        if (shouldPageElementsSkip() || !isPageBreak()) return this
        val textElement = this as Content.TextElement
        val rawLabel = textElement.segments.joinToString("") { it.text }.trim()
        if (rawLabel.isEmpty()) return this
        val label = pageLabel?.let { it(rawLabel) } ?: rawLabel
        return textElement.copy(
            segments =
                textElement.segments.mapIndexed { i, seg ->
                    if (i == 0) seg.copy(text = label) else seg.copy(text = "")
                },
        )
    }

    private fun Content.Element.isPageBreak(): Boolean {
        if (this !is Content.TextElement) return false
        val selector = locator.locations.cssSelector ?: return false
        return selector.startsWith("#") && selector.drop(1) in pageBreakIds
    }
}

File: flutter_readium/ios/flutter_readium/Sources/flutter_readium/navigator/PageBreakSkippingContentIteratorFactory.swift
import Foundation
import ReadiumShared

public enum PageBreakBehavior {
    /// Pass page-break elements through unchanged — the raw label text is spoken.
    case readAsIs
    /// Rewrite the label with a localized prefix, e.g. "Page 42" or "side 42".
    case prefixLabel
    /// Filter page-break elements out of the iterator entirely.
    case skip

    public init(from string: String?) {
        switch string?.lowercased() {
        case "readasis": self = .readAsIs
        case "prefixlabel": self = .prefixLabel
        case "skip": self = .skip
        default: self = .readAsIs
        }
    }
}

/// Wraps `HTMLResourceContentIterator.Factory` and handles EPUB page-break elements for TTS.
///
/// `pageBreakBehavior` controls how page-break elements are handled:
/// - `.readAsIs` (default) — element passes through unchanged.
/// - `.prefixLabel` — label text is rewritten to a localized string (e.g. "Page 42").
/// - `.skip` — element is filtered out entirely.
///
/// Page-break elements are identified by matching the element's CSS selector against
/// the fragment IDs declared in the publication's pageList nav.
final class PageBreakSkippingContentIteratorFactory: ResourceContentIteratorFactory {
    var pageBreakBehavior: PageBreakBehavior = .readAsIs

    private let delegate: ResourceContentIteratorFactory

    init(delegate: ResourceContentIteratorFactory = HTMLResourceContentIterator.Factory()) {
        self.delegate = delegate
    }

    func make(
        publication: Publication,
        readingOrderIndex: Int,
        resource: Resource,
        locator: Locator
    ) -> ContentIterator? {
        guard let inner = delegate.make(
            publication: publication,
            readingOrderIndex: readingOrderIndex,
            resource: resource,
            locator: locator
        ) else { return nil }

        let pageBreakIds = publication.manifest.pageBreakIds()
        guard !pageBreakIds.isEmpty else { return inner }

        return PageBreakHandlingIterator(
            delegate: inner,
            pageBreakIds: pageBreakIds,
            pageBreakBehavior: { [weak self] in self?.pageBreakBehavior ?? .readAsIs },
            pageLabel: publication.manifest.pageLabel()
        )
    }
}

private extension Manifest {
    func pageBreakIds() -> Set<String> {
        guard let links = subcollections["pageList"]?.first?.links else {
            return []
        }
        return Set(links.compactMap { link -> String? in
            guard let fragment = URL(string: link.href)?.fragment,
                  !fragment.isEmpty else { return nil }
            return fragment
        })
    }

    static let pageLabelFormats: [String: String] = [
        "en": "Page {label}",
        "da": "side {label}",
        "sv": "sida {label}",
        "no": "side {label}",
        "is": "{label}. síða",
    ]

    func pageLabel() -> ((String) -> String)? {
        guard let lang = metadata.languages.first else { return nil }
        let base = lang.components(separatedBy: "-").first?.lowercased() ?? lang.lowercased()
        guard let format = Self.pageLabelFormats[base] else { return nil }
        return { label in format.replacingOccurrences(of: "{label}", with: label) }
    }
}

private final class PageBreakHandlingIterator: ContentIterator {
    private let delegate: ContentIterator
    private let pageBreakIds: Set<String>
    private let pageBreakBehavior: () -> PageBreakBehavior
    private let pageLabel: ((String) -> String)?

    init(
        delegate: ContentIterator,
        pageBreakIds: Set<String>,
        pageBreakBehavior: @escaping () -> PageBreakBehavior,
        pageLabel: ((String) -> String)?
    ) {
        self.delegate = delegate
        self.pageBreakIds = pageBreakIds
        self.pageBreakBehavior = pageBreakBehavior
        self.pageLabel = pageLabel
    }

    func next() async throws -> ContentElement? {
        while let element = try await delegate.next() {
            if pageBreakBehavior() == .skip && isPageBreak(element) { continue }
            return transform(element)
        }
        return nil
    }

    func previous() async throws -> ContentElement? {
        while let element = try await delegate.previous() {
            if pageBreakBehavior() == .skip && isPageBreak(element) { continue }
            return transform(element)
        }
        return nil
    }

    private func transform(_ element: ContentElement) -> ContentElement {
        guard isPageBreak(element), pageBreakBehavior() == .prefixLabel,
              let textElement = element as? TextContentElement else { return element }
        guard let rawLabel = textElement.text?.trimmingCharacters(in: .whitespaces),
              !rawLabel.isEmpty else { return element }
        let label = pageLabel?(rawLabel) ?? rawLabel
        let newSegments = textElement.segments.enumerated().map { i, seg in
            TextContentElement.Segment(locator: seg.locator, text: i == 0 ? label : "", attributes: seg.attributes)
        }
        return TextContentElement(
            locator: textElement.locator,
            role: textElement.role,
            segments: newSegments,
            attributes: textElement.attributes
        )
    }

    private func isPageBreak(_ element: ContentElement) -> Bool {
        guard let textElement = element as? TextContentElement else { return false }
        guard let selector = textElement.locator.locations.cssSelector,
              selector.hasPrefix("#") else { return false }
        return pageBreakIds.contains(String(selector.dropFirst()))
    }
}
